File: ynot0117/121_Best_Time_to_buy_and_Sell_Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) { // 순서가 정해진 두 날짜 사이의 최대차이
        int max_profit = 0;
        int buy = 10001;

        for (auto price: prices)
        {
            if (price < buy)
                buy = price;
            if (max_profit < price - buy)
                max_profit = price - buy;
        }
        
        return max_profit;
    }
};

File: wjsdncl/121_best_time_to_buy_and_sell_stock.js
/*
/ 16분 소요
/
/ 시간복잡도: O(n)
/
/ 최소값과 최대 이익을 저장해두고 배열을 순회하며 최소값과 최대 이익을 갱신합니다.
/ 마지막 최대 이익을 반환합니다.
*/
var maxProfit = function (prices) {
  let minPrice = prices[0];
  let maxProfit = 0;

  // 배열 순회 (O(n))
  for (let i = 1; i < prices.length; i++) {
    // 최소값 갱신
    if (prices[i] < minPrice) {
      minPrice = prices[i];
    }
    // 최대 이익 갱신
    else if (prices[i] - minPrice > maxProfit) {
      maxProfit = prices[i] - minPrice;
    }
  }

  return maxProfit;
};
